### ISSUE TITLE
[Libretro] RPi4 DSP sound speed issue on v2.2

### ISSUE THREAD

**Issue by rtomasa** [flyinghead/flycast#1355]
**Platform / OS / Hardware**
Raspberry Pi OS Bookworm Lite (no desktop) 64bits running on RPi4

**Issue**
I'm creating a new libretro based frontend and so far I was using flycast with no issues for some time (ALSA via SDL2).

Today I decided to upgrade and compile the core to the latest stable tagged v2.2. Now I've noticed that when `reicast_enable_dsp = "enabled"` it makes the sound to play very fast for some time after loading games. Disabling the DSP fix the sounds speed issues but I had no problem in v2.1 and below.

BTW, the sound speed issues (and video too) happened also in previous versions when the core changed the framerate internally in some games. This could be amended by setting `reicast_detect_vsync_swap_interval = "enabled"`, but in the latest v2.2 this makes no difference at all, and only by setting `reicast_enable_dsp = "disabled"` the sound speed is working fine.

**Comment by flyinghead** [comment]
I don't see how enabling the DSP could make the sound play too fast. But it looks like you have some synchronization issue with your frontend.
The Flycast core relies on vsync to have "normal" framerate and sound output. Without it, it will run as fast as possible (fast forward).

**Comment by rtomasa** [comment]
It's weird because it happens in some games no matter the configuration. For example in Naomi2 VF4 the game goes very fast both a/v in very specific parts like the intros before the gameplay. Same happens in other games. I'm trying to figure out what's wrong because it is the only core facing this issues. I'm using DRM/KMS, and EGL for video and they use vsync by default. I will have a look and check again.

**Comment by flyinghead** [comment]
I've never heard anything like that. It sounds like the sampling rate is less than 44.1 kHz.
Is there some audio feature in the frontend that could explain this, like adaptive sampling rate?

Also I still have no clue why enabling dsp would trigger this. Flycast always outputs 44.1 kHz audio at all times.

**Comment by rtomasa** [comment]
I think that it is a mix of behaviours of my frontend + core. No idea what RetroArch makes to run it fine. For example, I saw that Retroarch is able to run Mslug6 at correct speed without enabling `reicast_detect_vsync_swap_interval`. In my case I need the core reports the change to 30fps to apply custom speed "hacks".

That said, I was doing more testing and changed my video engine to run using double buffered page flips instead of triple buffer, and now Crazy Taxi started to run fine with proper sound speed even with the DSP enabled, so I assume that the core is trying to measure some time between page flips to run both audio and video at the expected speed. I will continue doing more tests because it is not acceptable to me to have this core running with these inconsistencies, specially now that it runs flawlessly in Pi5.

**Comment by rtomasa** [comment]
@flyinghead After further investigation I found out that the issue is probably coming from my front-end implementation when dealing with cores using multithreading video. Same happens when running Mupen64 with video multithreading.
When I disable multithreading, sound is OK.

Any idea how should I deal with this scenario? I saw nothing special in the libretro API implementation, so not sure if I need to control the game speed in some special way for multithreading scenario.
